include/xpedite/common/WaitFreePointerEscrow.H

```diff
@@ -0,0 +1,100 @@
+//////////////////////////////////////////////////////////////////////////////////////////////////////
+//
+// WaitFreePointerEscrow - a SPSC wait free data structure for sharing a pointer between two threads.
+//
+// It has a clear ownership semantics, allowing the producer to deposit the pointer in the escrow,
+// and grant the consumer access to it.
+//
+// Once the consumer has successfully borrowed the pointer, the producer will know not to deallocate
+// the resource until the consumer has returned the lease.
+//
+// If the consumer takes too long to return the pointer, the producer can release the ownership. The scrow
+// will ensure the consumer knows it now owns the pointer and is responsible for deallocation.
+//
+// Author: Marcin Dlugajczyk, Morgan Stanley
+//
+//////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+#include <atomic>
+
+namespace xpedite { namespace common {
+
+  namespace {
+    static constexpr char borrowed{};
+  }
+
+  template<typename T>
+  class WaitFreePointerEscrow
+  {
+    std::atomic<T *> _ptr{nullptr};
+
+    // Returns a special pointer value used to indicate the pointer is currently borrowed.
+    constexpr T* getBorrowedMarker() {
+      return const_cast<T*>(reinterpret_cast<const T*>(&borrowed));
+    }
+
+  public:
+
+    WaitFreePointerEscrow() noexcept = default;
+    explicit WaitFreePointerEscrow(T *ptr_) noexcept : _ptr{ptr_} {}
+    WaitFreePointerEscrow<T>(WaitFreePointerEscrow<T> &other) = delete;
+    WaitFreePointerEscrow<T>(WaitFreePointerEscrow<T> &&other) noexcept = delete;
+    WaitFreePointerEscrow<T>& operator=(WaitFreePointerEscrow<T> &other) = delete;
+    WaitFreePointerEscrow<T>& operator=(WaitFreePointerEscrow<T> &&other) = delete;
+    ~WaitFreePointerEscrow() = default;
+
+    // Deposits the `ptr_` in the escrow and make it availabel to the consumer.
+    // Returns true if the pointer was susscessfully deposited in the escrow.
+    // Returns false if the escrow is already full.
+    bool deposit(T *ptr_) noexcept {
+      T *empty{nullptr};
+      return _ptr.compare_exchange_strong(empty, ptr_, std::memory_order_release, std::memory_order_acquire);
+    }
+
+    // Indicates if escrow holds a pointer.
+    bool empty() noexcept {
+      return _ptr.load(std::memory_order_acquire) == nullptr;
```

Comment:
This can be `std::memory_order_relaxed` as, checking whether the escrow is empty, doesn't need any ordering.